docs/config.md
# Configuration

For simplicity, mirakc uses a single YAML file for configuration.

Most of properties in the configuration are optional.  You can simply omit
properties which have default values listed below if the default values are
suitable for your environment.

| PROPERTY                         | DEFAULT                                   |
|----------------------------------|-------------------------------------------|
| [epg.cache-dir]                  | `None`                                    |
| [server.addrs]                   | `[{http: 'localhost:40772'}]`             |
| [server.workers]                 | The number of CPUs                        |
| [server.stream-chunk-size]       | `32768` (32KiB)                           |
| [server.stream-max-chunks]       | `1000`                                    |
| [server.stream-time-limit]       | `16000` (16s)                             |
| [channels\[\].name]              |                                           |
| [channels\[\].type]              |                                           |
| [channels\[\].channel]           |                                           |
| [channels\[\].extra-args]        | `''`                                      |
| [channels\[\].services]          | `[]`                                      |
| [channels\[\].excluded-services] | `[]`                                      |
| [channels\[\].disabled]          | `false`                                   |
| [tuners\[\].name]                |                                           |
| [tuners\[\].types]               |                                           |
| [tuners\[\].command]             |                                           |
| [tuners\[\].time-limit]          | `30000` (30s)                             |
| [tuners\[\].disabled]            | `false`                                   |
| [filters.pre-filter]             | `''`                                      |
| [filters.service-filter]         | `mirakc-arib filter-service --sid={{sid}}`|
| [filters.program-filter]         | `mirakc-arib filter-program --sid={{sid}} --eid={{eid}} --clock-pcr={{clock_pcr}} --clock-time={{clock_time}} --end-margin=2000 --pre-streaming` |
| [filters.post-filter]            | `''`                                      |
| [jobs.scan-services.command]     | `mirakc-arib scan-services{{#sids}} --sids={{.}}{{/sids}}{{#xsids}} --xsids={{.}}{{/xsids}}` |
| [jobs.scan-services.schedule]    | `'0 31 5 * * * *'` (execute at 05:31 every day) |
| [jobs.sync-clocks.command]       | `mirakc-arib sync-clocks{{#sids}} --sids={{.}}{{/sids}}{{#xsids}} --xsids={{.}}{{/xsids}}` |
| [jobs.sync-clocks.schedule]      | `'0 3 12 * * * *'` (execute at 12:03 every day) |
| [jobs.update-schedules.command]  | `mirakc-arib collect-eits{{#sids}} --sids={{.}}{{/sids}}{{#xsids}} --xsids={{.}}{{/xsids}}` |
| [jobs.update-schedules.schedule] | `'0 7,37 * * * * *'` (execute at 7 and 37 minutes every hour) |
| [mirakurun.openapi-json]         | `/etc/mirakurun.openapi.json`             |

[epg.cache-dir]: #epg.cache-dir
[server.addrs]: #server.addrs
[server.workers]: #server.workers
[server.stream-chunk-size]: #server.stream-chunk-size
[server.stream-max-chunks]: #server.stream-max-chunks
[server.stream-time-limit]: #server.stream-time-limit
[channels\[\].name]: #channels
[channels\[\].type]: #channels
[channels\[\].channel]: #channels
[channels\[\].extra-args]: #channels
[channels\[\].services]: #channels
[channels\[\].excluded-services]: #channels
[channels\[\].disabled]: #channels
[tuners\[\].name]: #tuners
[tuners\[\].types]: #tuners
[tuners\[\].command]: #tuners
[tuners\[\].time-limit]: #tuners
[tuners\[\].disabled]: #tuners
[filters.pre-filter]: #filters.pre-filter
[filters.service-filter]: #filters.service-filter
[filters.program-filter]: #filters.program-filter
[filters.post-filter]: #filters.post-filter
[jobs.scan-services.command]: #jobs.scan-services.command
[jobs.scan-services.schedule]: #jobs.scan-services.schedule
[jobs.sync-clocks.command]: #jobs.sync-clocks.command
[jobs.sync-clocks.schedule]: #jobs.sync-clocks.schedule
[jobs.update-schedules.command]: #jobs.update-schedules.command
[jobs.update-schedules.schedule]: #jobs.update-schdules.schedule
[mirakurun.openapi-json]: #mirakurun.openapi-json

## epg.cache-dir

An absolute path to a folder where EPG-related data will be stored.

`None` means that no data will be saved onto the filesystem.  In this case,
EPG-related data will be lost when mirakc stops.

```yaml
epg:
  cache-dir: /path/to/epg/cache
```

## server.addrs

`server.addrs` is a list of addresses to be bound.

There are two address types.

HTTP protocol:

```yaml
server:
  addrs:
    - http: '0.0.0.0:40772'
```

HTTPS protocol is not supported at this point.

UNIX domain socket:

```yaml
server:
  addrs:
    - unix: /var/run/mirakc.sock
```

mirakc never changes the ownership and permission of the socket.  Change them
after the socket has been created.  Or use SUID/SGID so that mirakc runs with
specific UID/GID.

Multiple addresses can be bound like below:

```yaml
server:
  addrs:
    - http: '0.0.0.0:40772'
    - unix: /var/run/mirakc.sock
```

## server.workers

The number of worker threads to serve the web API.

The specified number of threads are spawned and pooled at startup.

```yaml
server:
  workers: 2
```

## server.stream-chunk-size

The maximum size of a chunk for streaming.

```yaml
server:
  stream-chunk-size: 32768
```

An actual size of a chunk may be smaller than this value.

The default chunk size is 32 KiB which is large enough for 10 ms buffering.

## server.stream-max-chunks

The maximum number of chunks that can be buffered.

```yaml
server:
  stream-max-chunks: 1000
```

Chunks are dropped when the buffer is full.

The default maximum number of chunks is 1000 which is large enough for 10
seconds buffering if the chunk size is 32 KiB.

## server.stream-time-limit

The time limit for a streaming request.  The request will fail when the time
reached this limit.

```yaml
server:
  stream-time-limit: 20000
```

The value must be larger than `prepTime` defined in [EPGStation](https://github.com/l3tnun/EPGStation/blob/v1.6.9/src/server/Model/Operator/Recording/RecordingManageModel.ts#L45),
which is `15000` (15s) in v1.6.9.

This property is needed for avoiding the issue#1313 in actix-web in a streaming
request for a TV program.  In this case, no data is sent to the client until the
first TS packet comes from the streaming pipeline.  actix-web cannot detect the
client disconnect all that time due to the issue#1313.

## channels

Definitions of channels.  At least, one channel must be defined.

* name
  * An arbitrary name of the channel
* type
  * One of channel types in `GR`, `BS`, `CS` and `SKY`
* channel
  * A channel parameter used in a tuner command template
* extra-args
  * Extra arguments used in a tuner command template
* services
  * A list of SIDs (service identifiers) which must be included
  * An empty list means that all services found are included
* excluded-services
  * A list of SIDs which must be excluded
  * An empty list means that no service is excluded
  * Applied after processing the `services` property
* disabled (optional)
  * Disable the channel definition

```yaml
channels:
  - name: ETV
    type: GR
    channel: '26'

  # Disable NHK.
  - name: NHK
    type: GR
    channel: '27'
    disabled: true

  # Use only the service 101 in BS1.
  - name: BS1
    type: BS
    channel: BS15_0
    services: [101]

  # Exclude the service 531 from OUJ.
  - name: OUJ
    type: BS
    channel: BS11_2
    excluded-services: [531]

  # Extra arguments for szap-s2j
  - name: BS SPTV
    type: SKY
    channel: CH585
    extra-args: '-l JCSAT3A'
    serviceId: [33353]

  # Extra arguments for BonRecTest
  - name: ND02
    type: CS
    channel: '000'
    extra-args: '--space 1'
```

## tuners

Definitions of tuners.  At least, one tuner must be defined.

* name
  * An arbitrary name of the tuner
* types
  * A list of channel types supported by the tuner.
* command
  * A Mustache template string of a command to open the tuner
  * The command must output TS packets to `stdout`
* time-limit (optional)
  * A time limit in milliseconds
  * Stop streaming if no TS packet comes from the tuner for the time limit
* disabled (optional)
  * Disable the tuner

Command template variables:

* channel
  * The `channel` property of a channel defined in the `channels`
* channel_type
  * The `type` property of a channel defined in the `channels`
* duration
  * A duration to open the tuner in seconds
  * `-` means that the tuner is opened until the process terminates
  * TODO: `-` is always specified in the duration variable at this moment
* extra_args
  * The `extra-args` property of a channel defined in the `channels`

Cascading upstream Mirakurun-compatible servers is unsupported.  However, it's
possible to use upstream Mirakurun-compatible servers as tuners.  See the sample
below.

```yaml
tuners:
  - name: GR0
    types: [GR]
    command: recdvb {{channel}} {{duration}} -

  - name: Disabled
    types: [GR]
    command: cat /dev/null
    disabled: true

  # A tuner can be defined by using an "upstream" Mirakurun-compatible server.
  - name: upstream
    types: [GR, BS]
    command: >-
      curl -sG http://upstream:40772/api/channels/{{channel_type}}/{{channel}}/stream

```

## filters

Definitions of filters used in
[the streaming pipeline](./inside-mirakc.md#streaming-pipeline).

### filters.pre-filter

A Mustache template string of a command to process TS packets before
broadcasting the TS packets to subscribers.

The command must read TS packets from `stdin`, and output the processed TS
packets to `stdout`.

Template variables:

* index
  * The index of a tuner to be opened
* tuner_name
  * The name of a tuner identified by the index
* channel_name
  * The `name` property of a channel defined in the `channels`
* channel_type
  * The `type` property of a channel defined in the `channels`
* channel
  * The `channel` property of a channel defined in the `channels`

### filters.service-filter

A Mustache template string of a command to drop TS packets which are not
included in a service.

The command must read TS packets from `stdin`, and output the filtered TS
packets to `stdout`.

Template variables:

* sid
  * The idenfiter of the service

### filters.program-filter

A Mustache template string of a command to drop TS packets which are not
required for playback of a program in a service.

The command must read TS packets from `stdin`, and output the filtered TS
packets to `stdout`.

Template variables:

* sid
  * The idenfiter of the service
* eid
  * The event identifier of the program
* clock_pcr
  * A PCR value of synchronized clock for the service
* clock_time
  * A UNIX time (ms) of synchronized clock for the service

### filters.post-filter

A Mustache template string of a command to process TS packets after a packet
filter program.

The command must read TS packets from `stdin`, and output the processed TS
packets to `stdout`.

An empty string means that no post-filter command is applied even when the
`post-filter=true` query parameter is specified in a streaming API endpoint.

For compatibility with Mirakurun, the command is applied when the following both
conditions are met:

* The `decode=1` query parameter is specified
* The `post-filter` query parameter is NOT specified

Template variables:

* channel
  * The `channel` property of a channel defined in the `channels`
* channel_type
  * The `type` property of a channel defined in the `channels`
* sid (optional)
  * The idenfiter of the service
* eid (optional)
  * The event identifier of the program

## jobs

Definitions for background jobs.

Each job definition has the following two properties:

* command
  * A Mustache template string of a command
* schedule
  * A crontab expression of the job schedule
  * See https://crates.io/crates/cron for details of the format

### jobs.scan-services

The scan-services job scans audio/video services in channels defined in the
`channels`.

The command must read TS packets from `stdin`, and output the result to `stdout`

src/config.rs
use std::collections::HashMap;
use std::fs::File;
use std::sync::Arc;
use std::time::SystemTime;

use num_cpus;
use serde::Deserialize;
use serde_yaml;

use crate::models::{ChannelType, ServiceId};

pub fn load(config_path: &str) -> Arc<Config> {
    const SERVER_STREAM_TIME_LIMIT_MIN: u64 = 15_000;

    let reader = File::open(config_path)
        .unwrap_or_else(|err| {
            panic!("Failed to open {}: {}", config_path, err);
        });
    let mut config: Config = serde_yaml::from_reader(reader)
        .unwrap_or_else(|err| {
            panic!("Failed to parse {}: {}", config_path, err);
        });
    if config.server.stream_time_limit < SERVER_STREAM_TIME_LIMIT_MIN {
        log::warn!("server.stream_time_limit must be larger than {0}, \
                    reset it to {0}", SERVER_STREAM_TIME_LIMIT_MIN);
        config.server.stream_time_limit = SERVER_STREAM_TIME_LIMIT_MIN;
    }

    config.last_modified = std::fs::metadata(config_path)
        .map(|metadata| metadata.modified().ok()).ok().flatten();
    Arc::new(config)
}

// result

#[derive(Clone, Debug, Default, Deserialize, PartialEq)]
#[serde(rename_all = "kebab-case")]
#[serde(deny_unknown_fields)]
pub struct Config {
    #[serde(skip)]
    pub last_modified: Option<SystemTime>,
    #[serde(default)]
    pub epg: EpgConfig,
    #[serde(default)]
    pub server: ServerConfig,
    #[serde(default)]
    pub channels: Vec<ChannelConfig>,
    #[serde(default)]
    pub tuners: Vec<TunerConfig>,
    #[serde(default)]
    pub filters: FiltersConfig,
    #[serde(default)]
    pub pre_filters: HashMap<String, String>,
    #[serde(default)]
    pub post_filters: HashMap<String, String>,
    #[serde(default)]
    pub jobs: JobsConfig,
    #[serde(default)]
    pub recorder: RecorderConfig,
    #[serde(default)]
    pub mirakurun: MirakurunConfig,
}

#[derive(Clone, Debug, Default, Deserialize, PartialEq)]
#[serde(rename_all = "kebab-case")]
#[serde(deny_unknown_fields)]
pub struct EpgConfig {
    #[serde(default)]
    pub cache_dir: Option<String>,
}

#[derive(Clone, Debug, Deserialize, PartialEq)]
#[serde(rename_all = "kebab-case")]
#[serde(deny_unknown_fields)]
pub struct ServerConfig {
    #[serde(default = "ServerConfig::default_addrs")]
    pub addrs: Vec<ServerAddr>,
    #[serde(default = "ServerConfig::default_workers")]
    pub workers: usize,
    #[serde(default = "ServerConfig::default_stream_max_chunks")]
    pub stream_max_chunks: usize,
    #[serde(default = "ServerConfig::default_stream_chunk_size")]
    pub stream_chunk_size: usize,
    #[serde(default = "ServerConfig::default_stream_time_limit")]
    pub stream_time_limit: u64,
}

#[derive(Clone, Debug, Deserialize, PartialEq)]
#[serde(rename_all = "kebab-case")]
pub enum ServerAddr {
    Http(String),
    Unix(String),
}

impl ServerConfig {
    fn default_addrs() -> Vec<ServerAddr> {
        vec![ServerAddr::Http("localhost:40772".to_string())]
    }

    fn default_workers() -> usize {
        num_cpus::get()
    }

    fn default_stream_max_chunks() -> usize {
        // TODO: The same value is defined in the broadcaster module.
        1000
    }

    fn default_stream_chunk_size() -> usize {
        // TODO: The same value is defined in the broadcaster module.
        4096 * 8
    }

    fn default_stream_time_limit() -> u64 {
        16 * 1000  // 16s
    }
}

impl Default for ServerConfig {
    fn default() -> Self {
        ServerConfig {
            addrs: Self::default_addrs(),
            workers: Self::default_workers(),
            stream_max_chunks: Self::default_stream_max_chunks(),
            stream_chunk_size: Self::default_stream_chunk_size(),
            stream_time_limit: Self::default_stream_time_limit(),
        }
    }
}

#[derive(Clone, Debug, Deserialize, PartialEq)]
#[serde(rename_all = "kebab-case")]
#[serde(deny_unknown_fields)]
pub struct ChannelConfig {
    pub name: String,
    #[serde(rename = "type")]
    pub channel_type: ChannelType,
    pub channel: String,
    #[serde(default)]
    pub extra_args: String,
    #[serde(default)]
    pub services: Vec<ServiceId>,
    #[serde(default)]
    pub excluded_services: Vec<ServiceId>,
    #[serde(default)]
    pub disabled: bool,
}

#[derive(Clone, Debug, Deserialize, PartialEq)]
#[serde(rename_all = "kebab-case")]
#[serde(deny_unknown_fields)]
pub struct TunerConfig {
    pub name: String,
    #[serde(rename = "types")]
    pub channel_types: Vec<ChannelType>,
    pub command: String,
    #[serde(default = "TunerConfig::default_time_limit")]
    pub time_limit: u64,
    #[serde(default)]
    pub disabled: bool,
}

impl TunerConfig {
    fn default_time_limit() -> u64 {
        30 * 1000  // 30s
    }
}

#[derive(Clone, Debug, Deserialize, PartialEq)]
#[serde(rename_all = "kebab-case")]
#[serde(deny_unknown_fields)]
pub struct FiltersConfig {
    #[serde(default = "FiltersConfig::default_service_filter")]
    pub service_filter: String,
    #[serde(default = "FiltersConfig::default_program_filter")]
    pub program_filter: String,
    #[serde(default)]
    pub decode_filter: String,
}

impl FiltersConfig {
    fn default_service_filter() -> String {
        "mirakc-arib filter-service --sid={{sid}}".to_string()
    }

    fn default_program_filter() -> String {
        // The --pre-streaming option is NOT used anymore due to the issue#30.
        // Without the --pre-streaming option, actix-web cannot detect the
        // client disconnection until trying to write data to the socket.  In
        // the worst case, the client disconnect is not detected for a few
        // seconds which depends on a transmission interval of EIT p/f packets.
        //
        // The --pre-streaming was added in order to avoid the issue#1313 in
        // actix/actix-web.  PAT TS packets will be sent before the program
        // starts when this option is specified.  See masnagam/rust-case-studies
        // for details about the issue#1313.
        "mirakc-arib filter-program --sid={{sid}} --eid={{eid}} \
         --clock-pcr={{clock_pcr}} --clock-time={{clock_time}} \
         --end-margin=2000".to_string()
    }
}

impl Default for FiltersConfig {
    fn default() -> Self {
        FiltersConfig {
            service_filter: Self::default_service_filter(),
            decode_filter: String::new(),
            program_filter: Self::default_program_filter(),
        }
    }
}

#[derive(Clone, Debug, Deserialize, PartialEq)]
#[serde(rename_all = "kebab-case")]
#[serde(deny_unknown_fields)]
pub struct JobsConfig {
    #[serde(default = "JobsConfig::default_scan_services")]
    pub scan_services: JobConfig,
    #[serde(default = "JobsConfig::default_sync_clocks")]
    pub sync_clocks: JobConfig,
    #[serde(default = "JobsConfig::default_update_schedules")]
    pub update_schedules: JobConfig,
}

impl JobsConfig {
    fn default_scan_services() -> JobConfig {
        JobConfig {
            command: "mirakc-arib scan-services\
                      {{#sids}} --sids={{.}}{{/sids}}\
                      {{#xsids}} --xsids={{.}}{{/xsids}}".to_string(),
            schedule: "0 31 5 * * * *".to_string(),
        }
    }

    fn default_sync_clocks() -> JobConfig {
        JobConfig {
            command: "mirakc-arib sync-clocks\
                      {{#sids}} --sids={{.}}{{/sids}}\
                      {{#xsids}} --xsids={{.}}{{/xsids}}".to_string(),
            schedule: "0 3 12 * * * *".to_string(),
        }
    }

    fn default_update_schedules() -> JobConfig {
        JobConfig {
            command: "mirakc-arib collect-eits\
                      {{#sids}} --sids={{.}}{{/sids}}\
                      {{#xsids}} --xsids={{.}}{{/xsids}}".to_string(),
            schedule: "0 7,37 * * * * *".to_string(),
        }
    }
}

impl Default for JobsConfig {
    fn default() -> Self {
        JobsConfig {
            scan_services: Self::default_scan_services(),
            sync_clocks: Self::default_sync_clocks(),
            update_schedules: Self::default_update_schedules(),
        }
    }
}

#[derive(Clone, Debug, Deserialize, PartialEq)]
#[serde(rename_all = "kebab-case")]
#[serde(deny_unknown_fields)]
pub struct JobConfig {
    pub command: String,
    pub schedule: String,
}

#[derive(Clone, Debug, Deserialize, PartialEq)]
#[serde(rename_all = "kebab-case")]
#[serde(deny_unknown_fields)]
pub struct RecorderConfig {
    pub track_airtime_command: String,
}

impl Default for RecorderConfig {
    fn default() -> Self {
        RecorderConfig {
            track_airtime_command: "mirakc-arib track-airtime \
                                    --sid={{sid}} --eid={{eid}}".to_string(),
        }
    }
}

#[derive(Clone, Debug, Deserialize, PartialEq)]
#[serde(rename_all = "kebab-case")]
#[serde(deny_unknown_fields)]
pub struct MirakurunConfig {
    #[serde(default = "MirakurunConfig::default_openapi_json")]
    pub openapi_json: String,
}

impl MirakurunConfig {
    fn default_openapi_json() -> String {
        "/etc/mirakurun.openapi.json".to_string()
    }
}

impl Default for MirakurunConfig {
    fn default() -> Self {
        MirakurunConfig {
            openapi_json: Self::default_openapi_json(),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_config() {
        assert_eq!(
            serde_yaml::from_str::<Config>("{}").unwrap(),
            Default::default());

        let result = serde_yaml::from_str::<Config>(r#"
            unknown:
              property: value
        "#);
        assert!(result.is_err());
    }

    #[test]
    fn test_epg_config() {
        assert_eq!(
            serde_yaml::from_str::<EpgConfig>("{}").unwrap(),
            Default::default());

        assert_eq!(
            serde_yaml::from_str::<EpgConfig>(r#"
                cache-dir: /path/to/epg
            "#).unwrap(),
            EpgConfig {
                cache_dir: Some("/path/to/epg".to_string()),
            });

        let result = serde_yaml::from_str::<EpgConfig>(r#"
            unknown:
              property: value
        "#);
        assert!(result.is_err());
    }

    #[test]
    fn test_server_config() {
        assert_eq!(
            serde_yaml::from_str::<ServerConfig>("{}").unwrap(),
            Default::default());

        assert_eq!(
            serde_yaml::from_str::<ServerConfig>(r#"
                addrs:
                  - http: '0.0.0.0:40772'
            "#).unwrap(),
            ServerConfig {
                addrs: vec![
                    ServerAddr::Http("0.0.0.0:40772".to_string()),
                ],
                workers: ServerConfig::default_workers(),
                stream_max_chunks: ServerConfig::default_stream_max_chunks(),
                stream_chunk_size: ServerConfig::default_stream_chunk_size(),
                stream_time_limit: ServerConfig::default_stream_time_limit(),
            });

        assert_eq!(
            serde_yaml::from_str::<ServerConfig>(r#"
                addrs:
                  - unix: /path/to/sock
            "#).unwrap(),
            ServerConfig {
                addrs: vec![
                    ServerAddr::Unix("/path/to/sock".to_string()),
                ],
                workers: ServerConfig::default_workers(),
                stream_max_chunks: ServerConfig::default_stream_max_chunks(),
                stream_chunk_size: ServerConfig::default_stream_chunk_size(),
                stream_time_limit: ServerConfig::default_stream_time_limit(),
            });

        assert_eq!(
            serde_yaml::from_str::<ServerConfig>(r#"
                addrs:
                  - http: '0.0.0.0:40772'
                  - unix: /path/to/sock
            "#).unwrap(),
            ServerConfig {
                addrs: vec![
                    ServerAddr::Http("0.0.0.0:40772".to_string()),
                    ServerAddr::Unix("/path/to/sock".to_string()),
                ],
                workers: ServerConfig::default_workers(),
                stream_max_chunks: ServerConfig::default_stream_max_chunks(),
                stream_chunk_size: ServerConfig::default_stream_chunk_size(),
                stream_time_limit: ServerConfig::default_stream_time_limit(),
            });

        assert_eq!(
            serde_yaml::from_str::<ServerConfig>(r#"
                workers: 2
            "#).unwrap(),
            ServerConfig {
                addrs: ServerConfig::default_addrs(),
                workers: 2,
                stream_max_chunks: ServerConfig::default_stream_max_chunks(),
                stream_chunk_size: ServerConfig::default_stream_chunk_size(),
                stream_time_limit: ServerConfig::default_stream_time_limit(),
            });

        assert_eq!(
            serde_yaml::from_str::<ServerConfig>(r#"
                stream-max-chunks: 1000
            "#).unwrap(),
            ServerConfig {
                addrs: ServerConfig::default_addrs(),
                workers: ServerConfig::default_workers(),
                stream_max_chunks: 1000,
                stream_chunk_size: ServerConfig::default_stream_chunk_size(),
                stream_time_limit: ServerConfig::default_stream_time_limit(),
            });

        assert_eq!(
            serde_yaml::from_str::<ServerConfig>(r#"
                stream-chunk-size: 10000
            "#).unwrap(),
            ServerConfig {
                addrs: ServerConfig::default_addrs(),
                workers: ServerConfig::default_workers(),
                stream_max_chunks: ServerConfig::default_stream_max_chunks(),
                stream_chunk_size: 10000,
                stream_time_limit: ServerConfig::default_stream_time_limit(),
            });

        assert_eq!(
            serde_yaml::from_str::<ServerConfig>(r#"
                stream-time-limit: 10000
            "#).unwrap(),
            ServerConfig {
                addrs: ServerConfig::default_addrs(),
                workers: ServerConfig::default_workers(),
                stream_max_chunks: ServerConfig::default_stream_max_chunks(),
                stream_chunk_size: ServerConfig::default_stream_chunk_size(),
                stream_time_limit: 10000,
            });

        let result = serde_yaml::from_str::<ServerConfig>(r#"
            unknown:
              property: value
        "#);
        assert!(result.is_err());
    }

    #[test]
    fn test_channel_config() {
        assert!(serde_yaml::from_str::<ChannelConfig>("{}").is_err());

        assert_eq!(
            serde_yaml::from_str::<ChannelConfig>(r#"
                name: x
                type: GR
                channel: y
            "#).unwrap(),
            ChannelConfig {
                name: "x".to_string(),
                channel_type: ChannelType::GR,
                channel: "y".to_string(),
                extra_args: "".to_string(),
                services: vec![],
                excluded_services: vec![],
                disabled: false,
            });

        assert_eq!(
            serde_yaml::from_str::<ChannelConfig>(r#"
                name: x
                type: GR
                channel: y
                extra-args: "--extra args"
            "#).unwrap(),
            ChannelConfig {
                name: "x".to_string(),
                channel_type: ChannelType::GR,
                channel: "y".to_string(),
                extra_args: "--extra args".to_string(),
                services: vec![],
                excluded_services: vec![],
                disabled: false,
            });

        assert_eq!(
            serde_yaml::from_str::<ChannelConfig>(r#"
                name: x
                type: GR
                channel: y
                disabled: true
            "#).unwrap(),
            ChannelConfig {
                name: "x".to_string(),
                channel_type: ChannelType::GR,
                channel: "y".to_string(),
                extra_args: "".to_string(),
                services: vec![],
                excluded_services: vec![],
                disabled: true,
            });

        assert_eq!(
            serde_yaml::from_str::<ChannelConfig>(r#"
                name: x
                type: GR
                channel: y
                excluded-services: [100]
            "#).unwrap(),
            ChannelConfig {
                name: "x".to_string(),
                channel_type: ChannelType::GR,
                channel: "y".to_string(),
                extra_args: "".to_string(),
                services: vec![],
                excluded_services: vec![100.into()],
                disabled: false,
            });

        assert!(
            serde_yaml::from_str::<ChannelConfig>(r#"
                name: x
                type: WOWOW
                channel: y
            "#).is_err());

        let result = serde_yaml::from_str::<ChannelConfig>(r#"
            unknown:
              property: value
        "#);
        assert!(result.is_err());
    }

    #[test]
    fn test_tuner_config() {
        assert!(serde_yaml::from_str::<TunerConfig>("{}").is_err());

        assert_eq!(
            serde_yaml::from_str::<TunerConfig>(r#"
                name: x
                types: [GR, BS, CS, SKY]
                command: open tuner
            "#).unwrap(),
            TunerConfig {
                name: "x".to_string(),
                channel_types: vec![ChannelType::GR,
                                    ChannelType::BS,
                                    ChannelType::CS,
                                    ChannelType::SKY],
                command: "open tuner".to_string(),
                time_limit: TunerConfig::default_time_limit(),
                disabled: false,
            });

        assert_eq!(
            serde_yaml::from_str::<TunerConfig>(r#"
                name: x
                types: [GR, BS, CS, SKY]
                command: open tuner
                time-limit: 1
            "#).unwrap(),
            TunerConfig {
                name: "x".to_string(),
                channel_types: vec![ChannelType::GR,
                                    ChannelType::BS,
                                    ChannelType::CS,
                                    ChannelType::SKY],
                command: "open tuner".to_string(),
                time_limit: 1,
                disabled: false,
            });

        assert_eq!(
            serde_yaml::from_str::<TunerConfig>(r#"
                name: x
                types: [GR, BS, CS, SKY]
                command: open tuner
                disabled: true
            "#).unwrap(),
            TunerConfig {
                name: "x".to_string(),
                channel_types: vec![ChannelType::GR,
                                    ChannelType::BS,
                                    ChannelType::CS,
                                    ChannelType::SKY],
                command: "open tuner".to_string(),
                time_limit: TunerConfig::default_time_limit(),
                disabled: true,
            });

        assert!(
            serde_yaml::from_str::<TunerConfig>(r#"
                name: x
                types: [WOWOW]
                command: open tuner
            "#).is_err());

        let result = serde_yaml::from_str::<TunerConfig>(r#"
            unknown:
              property: value
        "#);
        assert!(result.is_err());
    }

    #[test]
    fn test_filters_config() {
        assert_eq!(
            serde_yaml::from_str::<FiltersConfig>("{}").unwrap(),
            Default::default());

        assert_eq!(
            serde_yaml::from_str::<FiltersConfig>(r#"
                service-filter: filter
            "#).unwrap(),
            FiltersConfig {
                service_filter: "filter".to_string(),
                decode_filter: "".to_string(),
                program_filter: FiltersConfig::default_program_filter(),
            });

        assert_eq!(
            serde_yaml::from_str::<FiltersConfig>(r#"
                decode-filter: filter
            "#).unwrap(),
            FiltersConfig {
                service_filter: FiltersConfig::default_service_filter(),
                decode_filter: "filter".to_string(),
                program_filter: FiltersConfig::default_program_filter(),
            });

        assert_eq!(
            serde_yaml::from_str::<FiltersConfig>(r#"
                program-filter: filter
            "#).unwrap(),
            FiltersConfig {
                service_filter: FiltersConfig::default_service_filter(),
                decode_filter: "".to_string(),
                program_filter: "filter".to_string(),
            });

        let result = serde_yaml::from_str::<FiltersConfig>(r#"
            unknown:
              property: value
        "#);
        assert!(result.is_err());
    }

    #[test]
    fn test_jobs_config() {
        assert_eq!(
            serde_yaml::from_str::<JobsConfig>("{}").unwrap(),
            Default::default());

        assert_eq!(
            serde_yaml::from_str::<JobsConfig>(r#"
                scan-services:
                  command: job
                  schedule: '*'
            "#).unwrap(),
            JobsConfig {
                scan_services: JobConfig {
                    command: "job".to_string(),
                    schedule: "*".to_string(),
                },
                sync_clocks: JobsConfig::default_sync_clocks(),
                update_schedules: JobsConfig::default_update_schedules(),
            });

        assert_eq!(
            serde_yaml::from_str::<JobsConfig>(r#"
                sync-clocks:
                  command: job
                  schedule: '*'
            "#).unwrap(),
            JobsConfig {
                scan_services: JobsConfig::default_scan_services(),
                sync_clocks: JobConfig {
                    command: "job".to_string(),
                    schedule: "*".to_string(),
                },
                update_schedules: JobsConfig::default_update_schedules(),
            });

        assert_eq!(
            serde_yaml::from_str::<JobsConfig>(r#"
                update-schedules:
                  command: job
                  schedule: '*'
            "#).unwrap(),
            JobsConfig {
                scan_services: JobsConfig::default_scan_services(),
                sync_clocks: JobsConfig::default_sync_clocks(),
                update_schedules: JobConfig {
                    command: "job".to_string(),
                    schedule: "*".to_string(),
                },
            });

        let result = serde_yaml::from_str::<JobsConfig>(r#"
            unknown:
              property: value
        "#);
        assert!(result.is_err());
    }

    #[test]
    fn test_job_config() {
        assert!(serde_yaml::from_str::<JobConfig>("{}").is_err());
        assert!(
            serde_yaml::from_str::<JobConfig>(r#"{"command":""}"#).is_err());
        assert!(
            serde_yaml::from_str::<JobConfig>(r#"{"schedule":""}"#).is_err());

        let result = serde_yaml::from_str::<JobConfig>(r#"
            unknown:
              property: value
        "#);
        assert!(result.is_err());
    }

    #[test]
    fn test_mirakurun_config() {
        assert_eq!(
            serde_yaml::from_str::<MirakurunConfig>("{}").unwrap(),
            Default::default());

        assert_eq!(
            serde_yaml::from_str::<MirakurunConfig>(r#"
                openapi-json: /path/to/json
            "#).unwrap(),
            MirakurunConfig {
                openapi_json: "/path/to/json".to_string(),
            });

        let result = serde_yaml::from_str::<MirakurunConfig>(r#"
            unknown:
              property: value
        "#);
        assert!(result.is_err());
    }
}
